### ISSUE TITLE
[replace] doesn't replace

### ISSUE THREAD

**Issue by P-E-Meunier** [NixOS/nixpkgs#22177]
## Issue description

The `[replace]` section of `Cargo.toml` is not handled by `buildRustPackage`, which issues the following warning (in my case, when replacing package `ring:0.6.2` with a local version):

```
warning: package replacement is not used: https://github.com/rust-lang/crates.io-index#ring:0.6.2
```

That section can be used to override a package with some locally fixed package.

This is not a big problem until one tries to quickly fix a bug in a dependency and `nixops` it into production.

### Steps to reproduce

Create a crate with the following `src/main.rs`:

```
extern crate ring;
fn main() {}
```

Clone `http://github.com/briansmith/ring` at the root of the crate.
And the following lines in `Cargo.toml`:

```
ring = "0.6.2"

[replace]
ring:0.6.2 = { path = "ring" }
```

Now package it with `rustBuildPackage`. The replacement is ignored.

## Technical details

* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
16.09.1515.89c567c (Flounder)

* Nix version: (run `nix-env --version`)
nix-env (Nix) 1.11.6

* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)
17.03pre-git

**Comment by wizeman** [comment]
I think this is not a bug in `rustBuildPackage`, but rather in your `Cargo.toml` or in `cargo`.

Are you sure you are specifying the version inside the `[replace]` section correctly?

You can check if the version is correct with `cargo pkgid ring`.

If you specified the version correctly, you can try instead to put the output of `cargo pkgid ring` as the package version to replace inside the `[replace]` section (as an `https://` link). This should work and get rid of the warning, according to https://github.com/servo/servo/issues/14688.

Alternatively, and this may require a lot more work, you can try to either switch to the unstable branch of nixpkgs/NixOS, or to backport all the Rust-related changes in the unstable branch to your own local git checkout of the stable branch.

I have done the latter some time ago and, I was able to successfully build the package as you specified in your "steps to reproduce" section (although, I had to add a `default.nix` file, add quotes around `ring:0.6.2`, and I also had to bump `ring` to version `0.6.3`, because `cargo` insisted I use that version, for some reason).

**Comment by wizeman** [comment]
It occurred to me that you could check a couple of additional issues:
1. Verify whether Nix is actually picking up any changes you have made to `Cargo.toml`, or the git checkout in `$project_dir/ring` - sometimes it's easy to forget to do a `git commit`, generate a tarball or whatever is necessary for Nix to pick up the changes.
2. From your description, it wasn't clear to me whether you tried to verify that you could build the package successfully without `rustBuildPackage`. If you can build it by manually doing `cargo fetch` and/or `cargo build`, that is another indication that the problem is not in `rustBuildPackage`.

**Comment by P-E-Meunier** [comment]
> Are you sure you are specifying the version inside the [replace] section correctly?

Yes, because `cargo build` is able to handle it correctly, in a `nix-shell` with the same versions of everything as my `rustBuildPackage`.

> you can try to either switch to the unstable branch of nixpkgs/NixOS, or to backport all the Rust-related changes in the unstable branch to your own local git checkout of the stable branch.

I'm working on a modified local checkout of nixpkgs, and I tried this both with my own rustup nightly fetcher, and the one in NixOS unstable.

My problem is, my project requires Rust nightly, and ring 0.6.2 stopped building on nightly this week (I could build it with rustup from January 22nd).

**Comment by BadDecisionsAlex** [comment]
I am having the same issue trying to make a package for "spotifyd"
Cargo fails as a result of multiple "rust-crypto" sources.
Did you ever find a solution?

**Comment by dmadisetti** [comment]
replace now crashes the build with a bad symlink
```txt
error: builder for '/nix/store/g5k2iq1gn2bw7c967dx2g6hjj2ahdil4-cargo-vendor-dir.drv' failed with exit code 1;
       last 1 log lines:
       > ln: failed to create symbolic link '/nix/store/dshpabbp5irzg2001fg5kj1spngb0260-cargo-vendor-dir/rnix-0.10.1/y4xbapipnj4vpp7b24srw43scnl5k62m-rnix-0.10.1': Permission denied
       For full logs, run 'nix log /nix/store/g5k2iq1gn2bw7c967dx2g6hjj2ahdil4-cargo-vendor-dir.drv'.
```

Full log is just the same line
